modules/afsmtp/afsmtp.c

```diff
@@ -119,51 +116,56 @@ afsmtp_dd_set_port(LogDriver *d, gint port)
 }
 
 void
-afsmtp_dd_set_subject(LogDriver *d, const gchar *subject)
+afsmtp_dd_set_subject(LogDriver *d, const gchar *subject, GlobalConfig *cfg)
 {
   AFSMTPDriver *self = (AFSMTPDriver *)d;
 
-  g_free(self->subject);
-  self->subject = g_strdup(subject);
+  log_template_unref(self->subject_tmpl);
+  self->subject_tmpl = log_template_new(cfg, "subject");
+  self->subject_tmpl->template = g_strdup(subject);
 }
```

Comment:
The patch bc1a252 looks great at first sight. I think it does leak a
LogTemplate reference though. template_content would return a reference
which is not freed anywhere.

PS: I know that the free/destructor part of the parser sucks, it's very
easy to make a mistake there.


-- 
Bazsi

On Mon, Jul 13, 2015 at 5:10 PM, Scheidler, Balázs <
balazs.scheidler@balabit.com> wrote:

>
>
> --
> Bazsi
>
> On Mon, Jul 6, 2015 at 12:01 PM, hkrist <notifications@github.com> wrote:
>
>> In modules/afsmtp/afsmtp.c
>> <https://github.com/balabit/syslog-ng/pull/419#discussion_r33920366>:
>>
>> >  {
>> >    AFSMTPDriver *self = (AFSMTPDriver *)d;
>> >
>> > -  g_free(self->subject);
>> > -  self->subject = g_strdup(subject);
>> > +  log_template_unref(self->subject_tmpl);
>> > +  self->subject_tmpl = log_template_new(cfg, "subject");
>> > +  self->subject_tmpl->template = g_strdup(subject);
>> >  }
>>
>> I poked into that member, because you suggested that we should get rid of
>> the string representations of the templated strings.
>>
>
> that string is managed by LogTemplate internally, it is where it saves the
> compiled template. if you call log_template_compile() it will save its
> input there. Isn't that enough?
>
> Also, you don't need to specify a name to log_template_new(), just pass a
> NULL. LogTemplate should not have a name, the only reason it does is that
> the same struct is used as both the template itself and a configuration
> object (e.g. template "foobar" in the confiuguration) which requires a
> name. It should be split, but as long as it isn't we don't need to use name
> anywhere, except for the configuration parser which parses the top-level
> template statement.
>
>
>> In my latest commit bc1a252
>> <https://github.com/balabit/syslog-ng/commit/bc1a25203612c60faefe5fe61265908f51e62568>
>> I used the template_content rule and it takes care of calling
>> log_template_new, log_template_compile and also takes care of error
>> handling, so if the log_template_compile function runs to an invalid
>> template, syslog-ng fails the startup.
>>
> I'll check that. Thanks,
>
>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/balabit/syslog-ng/pull/419/files#r33920366>.
>>
>
>